sdks/java/io/rrio/src/main/java/org/apache/beam/io/requestresponse/ThrottleWithoutExternalResource.java

```diff
@@ -17,41 +17,651 @@
+      }
+    }
+
+    /**
+     * Instantiates an initial {@link RestrictionTracker.IsBounded#BOUNDED} {@link OffsetRange}
+     * restriction from [-1, {@link List#size()}). Defaults to [-1, 0) for null {@link
+     * KV#getValue()} elements.
+     */
+    @GetInitialRestriction
+    public OffsetRange getInitialRange(@Element KV<Integer, List<RequestT>> element) {
+      int size = 0;
+      if (element.getValue() != null) {
+        size = element.getValue().size();
+      }
+      return OffsetRange.ofSize(size);
+    }
+
+    /** Instantiates an {@link OffsetRangeTracker} from an {@link OffsetRange} instance. */
+    @NewTracker
+    public RestrictionTracker<OffsetRange, Integer> newTracker(
+        @Restriction OffsetRange restriction) {
+      return new OffsetRangeTracker(restriction);
+    }
+
+    /** Simply returns the {@link OffsetRange} restriction. */
+    @TruncateRestriction
+    public RestrictionTracker.TruncateResult<OffsetRange> truncate(
+        @Restriction OffsetRange restriction) {
+      return new RestrictionTracker.TruncateResult<OffsetRange>() {
+        @Override
+        public ThrottleWithoutExternalResource.@Nullable OffsetRange getTruncatedRestriction() {
+          return restriction;
+        }
+      };
+    }
+
+    /**
+     * The {@link GetInitialWatermarkEstimatorState} initializes to this DoFn's output watermark to
+     * a negative infinity timestamp via {@link BoundedWindow#TIMESTAMP_MIN_VALUE}. The {@link
+     * Instant} returned by this method provides the runner the value is passes as an argument to
+     * this DoFn's {@link #newWatermarkEstimator}.
+     */
+    @GetInitialWatermarkEstimatorState
+    public Instant getInitialWatermarkState() {
+      return BoundedWindow.TIMESTAMP_MIN_VALUE;
+    }
+
+    /**
+     * This DoFn uses a {@link WatermarkEstimators.Manual} as its {@link NewWatermarkEstimator},
+     * instantiated from an {@link Instant}. The state argument in this method comes from the return
+     * of the {@link #getInitialWatermarkState}.
+     */
+    @NewWatermarkEstimator
+    public WatermarkEstimator<Instant> newWatermarkEstimator(
+        @WatermarkEstimatorState Instant state) {
+      return new WatermarkEstimators.Manual(state);
+    }
+
+    @ProcessElement
+    public void process(
+        @Element KV<Integer, List<RequestT>> element,
+        ManualWatermarkEstimator<Instant> estimator,
+        RestrictionTracker<OffsetRange, Integer> tracker,
+        MultiOutputReceiver receiver) {
+
+      int size = 0;
+      if (element.getValue() != null) {
+        size = element.getValue().size();
+      }
+
+      incIfPresent(inputElementsCounter, size);
+
+      if (element.getValue() == null || element.getValue().isEmpty()) {
+        return;
+      }
+
+      while (tracker.tryClaim(tracker.currentRestriction().getCurrent() + 1)) {
+        Instant nextEmittedTimestamp =
+            estimator.currentWatermark().plus(configuration.getMaximumRate().getInterval());
```

Comment:
Rather than using the watermark here, can we use state to store the next emitted timestamp? That has stronger synchronization guarantees in the case of splits (which may each track their own watermark) or other elements arriving with new restriction trackers. I'm wondering if this is where the issues you've seen around processcontinuation are coming from

Reply:
State parameters are not allowed with ProcessElement methods of Splittable DoFns: See ALLOWED_SPLITTABLE_PROCESS_ELEMENT_PARAMETERS https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java#L145-L158 vs ALLOWED_NON_SPLITTABLE_PROCESS_ELEMENT_PARAMETERS https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java#L127-L142. 

https://gist.github.com/damondouglas/d4606679ceab30446b0225716ac2b8fd for example throws 
```
Exception in thread "main" java.lang.IllegalArgumentException: StateNotCompatibleWithSDF, @ProcessElement process(ValueState, KV, ManualWatermarkEstimator, RestrictionTracker, OutputReceiver): Illegal parameter type: StateParameter{referent=StateDeclaration{id=some-state, field=private final org.apache.beam.sdk.state.StateSpec com.github.damondouglas.journal.StateNotCompatibleWithSDF.someStateSpec, stateType=org.apache.beam.sdk.state.ValueState<java.lang.Long>}, alwaysFetched=false}
	at org.apache.beam.sdk.transforms.reflect.DoFnSignatures$ErrorReporter.throwIllegalArgument(DoFnSignatures.java:2409)
	at org.apache.beam.sdk.transforms.reflect.DoFnSignatures.checkParameterOneOf(DoFnSignatures.java:1307)
	at org.apache.beam.sdk.transforms.reflect.DoFnSignatures.analyzeProcessElementMethod(DoFnSignatures.java:1276)
```

Reply:
Huh, didn't know that - I guess it makes sense because splittable and global state don't really go together (this is part of my concern here actually)